src/oic/utils/shelve_wrapper.py

```diff
@@ -16,31 +19,46 @@ def __len__(self):
         return db.__len__()
 
     def has_key(self, key):
+        if six.PY2 and not isinstance(key, six.string_types):
+            key = self._normalizeString(key)
         return key in self
 
     def __contains__(self, key):
+        if six.PY2 and not isinstance(key, six.string_types):
+            key = self._normalizeString(key)
         db = self._reopen_database()
         return db.__contains__(key)
 
     def get(self, key, default=None):
+        if six.PY2 and not isinstance(key, six.string_types):
+            key = self._normalizeString(key)
         db = self._reopen_database()
         return db.get(key, default)
 
     def __getitem__(self, key):
+        if six.PY2 and not isinstance(key, six.string_types):
+            key = self._normalizeString(key)
         db = self._reopen_database()
         return db.__getitem__(key)
 
     def __setitem__(self, key, value):
+        if six.PY2 and not isinstance(key, six.string_types):
+            key = self._normalizeString(key)
         db = self._reopen_database()
         db.__setitem__(key, value)
 
     def __delitem__(self, key):
+        if six.PY2 and not isinstance(key, six.string_types):
+            key = self._normalizeString(key)
         db = self._reopen_database()
         db.__delitem__(key)
 
     def _reopen_database(self):
         return shelve.open(self.filename, writeback=True)
 
+    def _normalizeString(self, key):
+        return unicodedata.normalize('NFKD', key).encode('ascii', 'ignore')
```

Comment:
This looks wrong for two reasons:

1) Open ID Connect Core explicitly forbids Unicode Normalization for string comparision
https://openid.net/specs/openid-connect-core-1_0.html#StringOps

2) Is encode('ascii') correct here? I would expect utf-8.

A malicious attacker could send unicode access tokens that degenerate to some totally different 7-bit ASCII string when treated with this. Not sure if that would be an issue though, but in theory this could lead to client_id confusion.

For that reason i would suggest to just try an encode 'utf-8' without normalization and just throw an exception if it fails.